spec/unit/login.spec.ts

```diff
@@ -57,3 +57,46 @@ describe("SSO login URL", function () {
         });
     });
 });
+
+describe("refreshToken", () => {
+    let client: TestClient;
+
+    beforeEach(function () {
+        client = new TestClient();
+    });
+
+    afterEach(function () {
+        client.stop();
+    });
+
+    it("requests the correctly-prefixed /refresh endpoint when server correctly accepts /v3", async () => {
+        const response = {
+            access_token: "access_token",
+            refresh_token: "refresh_token",
+            expires_in_ms: 30000,
+        };
+        client.httpBackend.when("POST", "/_matrix/client/v3/refresh").respond(200, response);
+        client.httpBackend.when("POST", "/_matrix/client/v1/refresh").check((_request) => {
+            fail("/v1/refresh unexpectedly called");
+        });
+
+        const refreshResultPromise = client.client.refreshToken("initial_refresh_token");
+        expect(await client.httpBackend.flush(undefined)).toEqual(1);
+        expect(await refreshResultPromise).toEqual(response);
+    });
+
+    it("falls back to /v1 when server does not recognized /v3 refresh", async () => {
+        const response = {
+            access_token: "access_token",
+            refresh_token: "refresh_token",
+            expires_in_ms: 30000,
+        };
+
+        client.httpBackend.when("POST", "/_matrix/client/v3/refresh").respond(400, { errcode: "M_UNRECOGNIZED" });
+        client.httpBackend.when("POST", "/_matrix/client/v1/refresh").respond(200, response);
+
+        const refreshResultPromise = client.client.refreshToken("initial_refresh_token");
+        await client.httpBackend.flushAllExpected();
```

Comment:
Is there a better way to assert "these two `ExpectedRequests` were received"?

Reply:
I suspect that, in view of my other comment, your question becomes moot, but to answer it: this is fine as it is: `flushAllExpected` will ensure that both of the expected calls have been made, so I don't think you need to do anything further.

A preferable solution IMHO is not to use `TestClient` (which comes with a load of baggage, and forces you to use the rather painful `matrix-mock-request`), and instead to instantiate a MatrixClient yourself and use `fetch-mock-jest` to intercept the HTTP calls. We're gradually trying to transition away from `TestClient` and `matrix-mock-request`.

Reply:
Switched to using `fetch-mock-jest` in [Add tests to ensure /v3/refresh is called + automatic /v1 retry](https://github.com/matrix-org/matrix-js-sdk/pull/3016/commits/036c0e03bc9b64ebad9ddaa1db629dab5f450520) – easy to use, thanks for the recommendation.